File: mobile/lib/src/core/ui/eposta.dart
/// (P233 §4) E-POSTA — biçim ve uzunluk, TEK kaynak.
///
/// Panel ikizi: `admin-web/lib/eposta.ts` (aynı kurallar, aynı kimlikler).
///
/// =========================================================================
/// ÖLÇÜLEN DURUM: sunucu ZATEN reddediyordu, kullanıcı NEDENİNİ görmüyordu
/// =========================================================================
/// Backend `EmailStr` kullanıyor ve ölçüldü: yerel kısım 64'ten uzunsa,
/// toplam 254'ten uzunsa ya da biçim bozuksa 422 dönüyor. Yani KURAL
/// VARDI. Eksik olan, kuralın KULLANICIYA SÖYLENMESİYDİ — alanların
/// hiçbirinde uzunluk sınırı ve alan-içi hata yoktu.
///
/// =========================================================================
/// NEDEN "SUNUCUYLA AYNI REGEX" DEĞİL
/// =========================================================================
/// İstemci doğrulaması sunucudan DAHA SIKI olursa gerçek bir adresi
/// reddeder ve kullanıcı kaydolamaz — geri dönüşü olmayan taraf budur. Bu
/// yüzden biçim denetimi KASITLI OLARAK GEVŞEK: yalnızca hiçbir sunucunun
/// kabul etmeyeceği şeyleri reddeder. Kesin karar sunucunun.
library;

/// RFC 5321: yerel kısım en çok 64 sekizli.
const kEpostaYerelSinir = 64;

/// RFC 5321: adresin tamamı en çok 254 karakter.
const kEpostaSinir = 254;

/// Doğrulama sonucu — METİN DEĞİL KİMLİK (README §15).
enum EpostaHatasi { bos, bicim, yerelUzun, cokUzun }

/// [ham] için hata kimliği; `null` = geçerli.
EpostaHatasi? epostaHatasi(String ham, {bool zorunlu = true}) {
  final s = ham.trim();
  if (s.isEmpty) return zorunlu ? EpostaHatasi.bos : null;
  // UZUNLUK ÖNCE: biçim denetimi uzun bir adreste de geçebilir ve kullanıcı
  // asıl engeli (uzunluk) hiç görmezdi.
  if (s.length > kEpostaSinir) return EpostaHatasi.cokUzun;
  final parcalar = s.split('@');
  if (parcalar.length != 2) return EpostaHatasi.bicim;
  final yerel = parcalar[0];
  final alan = parcalar[1];
  if (yerel.length > kEpostaYerelSinir) return EpostaHatasi.yerelUzun;
  if (yerel.isEmpty || alan.isEmpty) return EpostaHatasi.bicim;
  if (RegExp(r'\s').hasMatch(s)) return EpostaHatasi.bicim;
  if (!alan.contains('.')) return EpostaHatasi.bicim;
  if (alan.startsWith('.') || alan.endsWith('.')) return EpostaHatasi.bicim;
  if (yerel.startsWith('.') || yerel.endsWith('.')) return EpostaHatasi.bicim;
  if (s.contains('..')) return EpostaHatasi.bicim;
  return null;
}

/// Saklanacak değer: kırpılmış + küçük harf.
String epostaNormalle(String ham) => ham.trim().toLowerCase();

File: admin-web/lib/eposta.ts
import { SINIR } from "@/lib/girdi-siniri";

/**
 * (P233 §4) E-POSTA — bicim ve uzunluk, TEK kaynak.
 *
 * =========================================================================
 * OLCULEN DURUM: sunucu ZATEN reddediyordu, kullanici NEDENINI gormuyordu
 * =========================================================================
 * Backend `EmailStr` (pydantic/email-validator) kullaniyor ve OLCULDU:
 * yerel kisim 64'ten uzunsa, toplam 254'ten uzunsa ya da bicim bozuksa
 * 422 donuyor. Yani KURAL VARDI. Eksik olan, kuralin KULLANICIYA
 * SOYLENMESIYDI: alanlarin hicbirinde `maxLength` ve alan-ici hata yoktu
 * (6 web alani, 4 mobil alan tarandi); kullanici formu doldurup
 * gonderdikten SONRA jenerik bir hata goruyordu.
 *
 * =========================================================================
 * NEDEN "SUNUCUYLA AYNI REGEX" DEGIL
 * =========================================================================
 * Istemci dogrulamasi sunucudan DAHA SIKI olursa gercek bir adresi
 * reddeder ve kullanici kaydolamaz — geri donusu olmayan taraf budur.
 * Bu yuzden buradaki bicim denetimi KASITLI OLARAK GEVSEK: yalnizca
 * hicbir sunucunun kabul etmeyecegi seyleri (bosluk, `@` yoklugu/ikizi,
 * noktasiz alan adi, bas/son nokta, cift nokta) reddeder. Kesin karar
 * sunucunun.
 *
 * UZUNLUK SINIRLARI ise KESIN: RFC 5321 yerel kisim 64, toplam 254.
 */

/** RFC 5321: yerel kisim en cok 64 sekizli. */
export const EPOSTA_YEREL_SINIR = 64;

/** RFC 5321: adresin tamami en cok 254 karakter. (P248 §3a) Deger
 * sunucuyla ortak tek kaynaktan (`SINIR.EPOSTA`, kilit: girdi-siniri.test). */
export const EPOSTA_SINIR = SINIR.EPOSTA;

export type EpostaHatasi = "bos" | "bicim" | "yerelUzun" | "cokUzun";

/** [ham] icin hata kimligi; `null` = gecerli. Cumle CIZIM katmaninda. */
export function epostaHatasi(
  ham: string,
  zorunlu = true,
): EpostaHatasi | null {
  const s = (ham ?? "").trim();
  if (!s) return zorunlu ? "bos" : null;
  // UZUNLUK ONCE: bicim denetimi uzun bir adreste de gecebilir ve
  // kullanici asil engeli (uzunluk) hic gormezdi.
  if (s.length > EPOSTA_SINIR) return "cokUzun";
  const parcalar = s.split("@");
  if (parcalar.length !== 2) return "bicim";
  const [yerel, alan] = parcalar;
  if (yerel.length > EPOSTA_YEREL_SINIR) return "yerelUzun";
  if (!yerel || !alan) return "bicim";
  if (/\s/.test(s)) return "bicim";
  if (!alan.includes(".")) return "bicim";
  if (alan.startsWith(".") || alan.endsWith(".")) return "bicim";
  if (yerel.startsWith(".") || yerel.endsWith(".")) return "bicim";
  if (s.includes("..")) return "bicim";
  return null;
}

/** Saklanacak deger: kirpilmis + kucuk harf. */
export function epostaNormalle(ham: string): string {
  return (ham ?? "").trim().toLowerCase();
}
